### PR TITLE
[google_sign_in_web] Fix race condition on init.

### PR DESCRIPTION
**Pull request by ditman**
## Description

The Auth object that `.init()` returns may not be fully ready by the time we start calling methods on it; however it has a `.then()` method that gets called when it is *really* ready to go.

* Add a completer to signal when Auth2 is ready.
* Move auth2-dependent init to the `then` method.
  * onSuccess, complete the auth2 ready future.
  * onError, throw a PlatformException. This is normally triggered when the domain:port of the web app hasn't been whitelisted on the google sign in console, and things like that. Useful for debugging.

## Related Issues

* https://github.com/flutter/flutter/issues/48527
* https://github.com/flutter/flutter/issues/47832

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart [hunk 1]
```diff
@@ -81,19 +81,25 @@ class GoogleSignInPlugin extends GoogleSignInPlatform {
...
-      if (!initUserCompleter.isCompleted) {
-        initUserCompleter.complete(nextUser);
-      } else {
-        _lastSeenUser = nextUser;
-      }
+    Completer<void> isAuthInitialized = Completer<void>();
+    _isAuthInitialized = isAuthInitialized.future;
+
+    auth.then(allowInterop((auth2.GoogleAuth initializedAuth) {
+      // onSuccess
```
Comment (CaseyHillers) [inline packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart]: Thanks for adding this. I didn't realize this was from generated code so it doesn't have named parameters.

Comment (ditman) [reply packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart]: Yeah, I think there's a good opportunity for improvement here (auto-wrapping functions in allowInterop, and maybe named parameters).

I'll float the idea around here!

File: packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart [hunk 2]
```diff
@@ -81,19 +81,25 @@ class GoogleSignInPlugin extends GoogleSignInPlatform {
...
+
+      // TODO: https://github.com/flutter/flutter/issues/48528
+      // This plugin doesn't notify the app of external changes to the
+      // state of the authentication, i.e: if you logout elsewhere...
+
+      isAuthInitialized.complete();
+    }), allowInterop((dynamic reason) {
+      // onError
+      throw (PlatformException(
```
Comment (hterkelsen) [inline packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart]: nit: unnecessary parentheses

File: packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart [hunk 3]
```diff
@@ -39,14 +38,15 @@ class GoogleSignInPlugin extends GoogleSignInPlatform {
   }
 
   Future<void> _isGapiInitialized;
+  Future<void> _isAuthInitialized = Completer<void>().future;
```
Comment (hterkelsen) [inline packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart]: seems odd to set `_isAuthInitialized` to a `Future` that will never complete. If someone does `await initialized;` before `init()` has been called, then they will be awaiting forever.

I suggest creating the `Completer<void>` for `_isAuthInitialized` in the constructor

Comment (ditman) [reply packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart]: I was thinking about this. I don't like futures that never complete either.

I'll add some sort of boolean flag that I can quickly check (and throw a StateError or similar) when you call other methods before init, and THEN, have this as a proper completer that will be created at the right moment by the .init function.

That way the plugin can bail out early, if you call anything before init, and will be able to wait for init to be ready elsewhere.

Thanks for the nudge :P

File: packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart (hunk 2 above)
Comment (ditman) [reply packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart]: Totally forgot about these. I just removed them!

File: packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart (hunk 1 above)
Comment (ditman) [reply packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart]: I cut a ticket to the tool that we used to generate the code:

* https://github.com/dart-lang/js_facade_gen/issues/78

_(BTW, we can contribute to the tool very easily, WINK, WINK)_

**Comment by ditman** [comment]
Will merge and publish on green. In addition to unit tests, I tested this manually with the example app, and the Cocoon project.

**Comment by ditman** [comment]
[Published](https://pub.dev/packages/google_sign_in_web#083) and [tagged](https://github.com/flutter/plugins/commits/google_sign_in_web-v0.8.3) as v0.8.3.

**Pull request merged**